tools/webdriver/webdriver/bidi/client.py

```diff
@@ -0,0 +1,393 @@
+            if future is None:
+                raise ValueError(f"No pending command with id {data['id']}")
+            if "result" in data:
+                future.set_result(data["result"])
+            elif "error" in data and "message" in data:
+                assert isinstance(data["error"], str)
+                assert isinstance(data["message"], str)
+                future.set_exception(BidiException(data["error"],
+                                                   data["message"],
+                                                   data.get("stacktrace")))
+            else:
+                raise ValueError(f"Unexpected message: {data!r}")
+        elif "method" in data and "params" in data:
+            # This is an event
+            method = data["method"]
+            listeners = self.event_listeners.get(method, [])
+            if not listeners:
+                listeners = self.event_listeners.get(None, [])
+            for listener in listeners:
+                await listener(method, data["params"])
+        else:
+            raise ValueError(f"Unexpected message: {data!r}")
+
+    async def end(self) -> None:
+        """Close websocket connection."""
+        assert self.transport is not None
+        await self.transport.end()
+
+    def add_event_listener(self,
+                           name: Optional[str],
+                           fn: Callable[[str, Mapping[str, Any]], Awaitable[Any]]) -> None:
+        """Add a listener for the event with a given name.
+
+        If name is None, the listener is called for all messages that are not otherwise
+        handled.
+
+        :param name: Name of event to listen for or None to register a default handler
+        :param fn: Async callback function that receives event data
+        """
+        self.event_listeners[name].append(fn)
+
+
+class Transport:
+    """Low level message handler for the WebSockets connection"""
+    def __init__(self, url: str,
+                 msg_handler: Callable[[Mapping[str, Any]], Coroutine[Any, Any, None]],
+                 loop: Optional[asyncio.AbstractEventLoop] = None):
+        self.url = url
+        self.connection: Optional[websockets.WebSocketClientProtocol] = None
+        self.msg_handler = msg_handler
+        self.send_buf: List[Mapping[str, Any]] = []
+
+        if loop is None:
+            loop = get_running_loop()
+        self.loop = loop
+
+        self.read_message_task: Optional[asyncio.Task[Any]] = None
+
+    async def start(self) -> None:
+        self.connection = await websockets.client.connect(self.url)
+        self.read_message_task = self.loop.create_task(self.read_messages())
+
+        for msg in self.send_buf:
+            await self._send(self.connection, msg)
+
+    async def send(self, data: Mapping[str, Any]) -> None:
+        if self.connection is not None:
+            await self._send(self.connection, data)
+        else:
+            self.send_buf.append(data)
+
+    @staticmethod
```

Comment:
What's the benefit by having it as a static method?

Reply:
I think I did this to make the type checking easier since in this method you have to have a valid connection (but we don't depend on self anywhere). But it's probably not important to factor this out of `send`.